File: Longest Substring Without Repeating Characters - Leetcode 3/brute force.cpp
#include <unordered_set>
#include <algorithm>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(const string s) {

        int longest = 0;

        for (auto i = 0; i < s.size(); ++i) {
            for (auto substr_len = 1; i + substr_len <= s.size(); ++substr_len) {
                unordered_set<char> seen{s.cbegin() + i, s.cbegin() + i + substr_len};
                if (seen.size() == substr_len) 
                    longest = max(longest, substr_len);
            }
        }

        return longest;
    }
};
// Time Complexity: O(n^3)
// Space Complexity: O(n)

File: Longest Substring Without Repeating Characters - Leetcode 3/brute force.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        max_length = -1
        for i in range(n):
            for substr_len in range(1, n - i + 1):
                substr = s[i: i+substr_len]
                seen = set(substr) # O(m) where m = len(substr)
                if substr_len == len(seen):
                    # there are no repeated characters in the substr
                    max_length = max(max_length, substr_len)

        return max_length

# Time Complexity: O(n^3)
# Space Complexity: O(n)
